### ISSUE TITLE
Wibar has a frozen/copied buffer after waking from sleep

### ISSUE THREAD

**Issue by microparsec** [awesomeWM/awesome#2470]
Output of `awesome --version`:

awesome v4.2 (Human after all)
 • Compiled against Lua 5.3.4 (running with Lua 5.3)
 • D-Bus support: ✔
 • execinfo support: ✔
 • xcb-randr version: 1.5
 • LGI version: 0.9.2

**How to reproduce the issue:**

I use this code in `rc.lua` to create two wibars on each of my screen.

    awful.screen.connect_for_each_screen(function(s)
        -- Wallpaper
        set_wallpaper(s)

        -- Each screen has its own tag table.
        awful.tag({ "HOME", "WWW", "MEDIA", "DEV", "SYS", "CONF" }, s, awful.layout.layouts[1])

        -- Create a promptbox for each screen
        s.mypromptbox = awful.widget.prompt()
        -- Create an imagebox widget which will contain an icon indicating which layout we're using.
        -- We need one layoutbox per screen.
        s.mylayoutbox = awful.widget.layoutbox(s)
        s.mylayoutbox:buttons(gears.table.join(
                               awful.button({ }, 1, function () awful.layout.inc( 1) end),
                               awful.button({ }, 3, function () awful.layout.inc(-1) end),
                               awful.button({ }, 4, function () awful.layout.inc( 1) end),
                               awful.button({ }, 5, function () awful.layout.inc(-1) end)))
        -- Create a taglist widget
        s.mytaglist = awful.widget.taglist(s, awful.widget.taglist.filter.all, taglist_buttons)

        -- Create a tasklist widget
        s.mytasklist = awful.widget.tasklist(s, awful.widget.tasklist.filter.currenttags, tasklist_buttons)

        -- Create the wibox
        s.mywibox = awful.wibar({ position = "top", screen = s })
        s.taskbarwibox = awful.wibar({ position = "bottom", screen = s})

        s.taskbarwibox:setup {
	    layout = wibox.layout.align.horizontal,
	    {
	        layout = wibox.layout.align.horizontal,
	        s.mytasklist
            },
	    nil,
	    {
	        layout = wibox.layout.align.horizontal,
	        nil
    	    }
        }

        -- Add widgets to the wibox
        s.mywibox:setup {
            layout = wibox.layout.align.horizontal,
            { -- Left widgets
                layout = wibox.layout.fixed.horizontal,
                s.mytaglist,
	        mymediaplayer,
	        spacer,
                s.mypromptbox,
            },
            nil, -- Middle widget
            { -- Right widgets
                layout = wibox.layout.fixed.horizontal,
                mykeyboardlayout,
                wibox.widget.systray(),
	        spacer,
	        mycpuicon,
	        mycpupercentage,
	        spacer,
	        mybatterystatus,
	        spacer,
                mytextclock,
                s.mylayoutbox,
            },
        }
    end)
    -- }}}

The top one contains the screen buttons, a media interface, and on the right the status indicators.
My bottom wibar contains the current windows of the current screen, nothing else.

***The problem:***
After waking from sleep, the status (right) part of the top wibar is also shown on the bottom wibar, even thought nothing is programmed to be displayed there. Reloading Awesome resolves the issue. Opening enough windows so that the taskbar in the bottom wibar reaches the right of the screen, overwrites the pixels in the bottom wibar. 

I will experiment with placement of different elements and see if I can find the breaking point. In the meantime I hope to have illustrated the issue I'm having here clearly. Any help or suggestions are appreciated. 

**Actual result:**

![Screenshot of unusual behaviour](https://i.imgur.com/5WqbOrR.jpg)

**Expected result:**

There shouldn't be a copy of the

 status indicators in the bottom wibar

**Comment by psychon** [comment]
Uhm... okay.
Could you run the following in a terminal and check if the "strange buffer copy" is also visible in `/tmp/t.png`?
```
awesome-client 'return require("gears.surface").load_silently(screen[1].taskbarwibox._drawable.drawable.surface, false):write_to_png("/tmp/t.png")'
```
Also, could you check if running the following makes "the things" go away? It, well, forces a complete repaint.
```
awesome-client 'return screen[1].taskbarwibox._drawable._do_complete_repaint()'
```
(Replace `screen[1]` with the appropriate number, if needed)

**Comment by psychon** [comment]
Oh and: Does `s.taskbarwibox:set_widget(s.mytasklist)` perhaps make a difference? (Or at least: Getting rid of one of the many `wibox.layout.align.horizontal` instances)

**Comment by microparsec** [comment]
> Uhm... okay.
> Also, could you check if running the following makes "the things" go away? It, well, forces a complete repaint.
> 
> (Replace `screen[1]` with the appropriate number, if needed)

Yes, the repaint makes the things go away.

> Oh and: Does `s.taskbarwibox:set_widget(s.mytasklist)` perhaps make a difference? (Or at least: Getting rid of one of the many `wibox.layout.align.horizontal` instances)

The original didn't have as many wibox.layout... to be honest, I added them just to `nil` them. For explicit emptyness, so to say. Originally it read:

    s.taskbarwibox:setup {
        layout = wibox.layout.align.horizontal,
        s.mytasklist
    }

However, your suggestion to use "set_widget" seems to have resolved the issue. Also, the tasklist now spans the entire wibar, instead of only using the minimal width of the wibar. (That is actually not completely what I want, but I will just tinker a bit more with it.)

Next thing I instincively did was remove the tasklist form the bottom wibar completely, so that it would just be an empty bar. After cycling sleep a few times to test, no signes of the buffer copy now either. 

So my conclusion is that this behaviour must come from within the wibar.layout instances I used previously. Maybe because they didn't have unique id's, as @actionless suggested, and therefore copied eachothers buffers for a moment before _not_ getting updated to "empty". That comes with a sidenote that I actually don't know how awesome works at all. 

I hope you can get something useful out of this thread. If you have no further investigation ongoing or questions/suggestions you'd like me to try, please close and mark this thread as solved.

**Comment by psychon** [comment]
Could you test if running awesome with the `--no-argb` command line option makes the problem also go away?

(To be precise: Could you un-do your changes to your config, test that the issue appears again after wake-up-from-sleep, and then test that starting awesome with `--no-argb` makes the problem go away?)

The reason I ask is that similar symptoms were reported in #2590 and there `--no-argb` made the issue go away.

**Comment by microparsec** [comment]
awesome v4.2 (Human after all)
 • Compiled against Lua 5.3.5 (running with Lua 5.3)
 • D-Bus support: ✔
 • execinfo support: ✔
 • xcb-randr version: 1.6
 • LGI version: 0.9.2

After reverting the changes and reproducing the bug, adding the --no-argb flag made no difference, except for disabling my beautiful transparent effects.

I want to retcon a bit of my issue. It turns out the status bar isn't duplicated on the bottom bar until I switch screen/workspace. So when I am in the MEDIA screen, right after wake, everything is fine. It is only after I switch to WWW that the right portion of the top bar is duplicated in the right portion of the bottom bar.

I have also added unique id's to every layout, as to hopefully fix this issue as suggested earlier.

There are even more problems with sleeping and waking. Sometimes the entire top bar dissapears except for the screenbuttons, and when the clock ticks a minute the status part reappears as well. When pressing Super+R the command launcher appears too. This issue is however far more sporadic, and only happens (a guesstimate) 1 out of every 15 wakes. I have no clue if it is related, but it might as well be.

I have uploaded my rc.lua, theme and lain plugins in the included zipfile, if anyone is willing to try reproduce the issue with my settings, their feedback is greatly appreciated!
[spectre.zip](https://github.com/awesomeWM/awesome/files/2768678/spectre.zip)

**Comment by actionless** [comment]
could it be what while handling signals of screen appearing/disappearing something made delayed_call() so event loop freed and caught next signal for screen appear/disappear before fully finished handling previous?

particularly, can you reproduce it on single screen?

**Comment by actionless** [comment]
also, is the second panel functional? or it's just a redraw glitch on top of your wallpaper?

**Comment by actionless** [comment]
does that mean what you can't reproduce the problem with the default config?

if so, do you have any delayed_call or async function calls in your config?

**Comment by microparsec** [comment]
I can reproduce the problem with the default config. I have included my edited rc.lua in the attachment below. The diff is the following:

```
206a207,215
>     -- Create the bottom taskbar
>     s.taskbar = awful.wibar({ position = "bottom", screen = s })
>     s.taskbar:setup{
> 	layout = wibox.layout.align.horizontal,
> 	
> 	    s.mytasklist
> 	
>     }
> 
216c225
<         s.mytasklist, -- Middle widget
---
>         nil, -- Middle widget
```
I moved the s.mytasklist to the new wibar that is positioned on the bottom of each screen.

After waking and switching screen the clock etc. is duplicated on the bottom wibar:

![Screenshot of buffer glitch on default rc.lua](https://i.imgur.com/cBUqnrg.png)

I have also managed the capture of the worse glitch, where whole parts seem to disappear:

![Screenshot of even worse wake after sleep glitch](https://i.imgur.com/biIAS14.jpg)

Attachment:
[Default rc lua with bottom taskbar: rc.lua](https://github.com/awesomeWM/awesome/files/2770443/rc.lua.changemyextension.txt)

**Comment by actionless** [comment]
and with set_widget your minimal example on default config works always fine?
```
>     s.taskbar = awful.wibar({ position = "bottom", screen = s })
>     local layout = wibox.layout.align.horizontal()
>     layout:set_first(s.mytasklist)
>     s.taskbar:set_widget(layout)
```

while adding id's not making any difference?

**Comment by microparsec** [comment]
using set_widget doesn't seem to make a difference, and neither does giving the wibars unique id's:

![Screenshot of glitch in standard rc.lua using set_widget](https://i.imgur.com/sobQpqr.png)

**Comment by actionless** [comment]
i think it could be useful to log the fired signals which could give some hint what's going on

@blueyed was working on something like that: https://github.com/awesomeWM/awesome/pull/1706/files
but i am not sure if the patch is working (wasn't testing myself)

**Comment by microparsec** [comment]
Right, that is some more hardcore stuff. I have never really looked into the inner working of programs of this scale. 
I don't know if I have time to get into it this coming week, so we will have to put the troubleshooting on hold for a while. Fear not though, I will return with results in the near future. 
I will try to replicate the problem on another machine too, to hopefully eliminate the possibility of this being a hardware specific issue.
